packages/keymanager-server/src/impl.ts

```diff
@@ -0,0 +1,207 @@
+        await fs.promises.writeFile(keystorePath, keystoreStr, {encoding: "utf8"});
+        lockfile.lockSync(lockFilePath);
+
+        statuses[i] = {status: ImportStatus.imported};
+      } catch (e) {
+        statuses[i] = {status: ImportStatus.error, message: (e as Error).message};
+      }
+    }
+
+    return {data: statuses};
+  }
+
+  /**
+   * DELETE must delete all keys from `request.pubkeys` that are known to the keymanager and exist in its
+   * persistent storage. Additionally, DELETE must fetch the slashing protection data for the requested keys from
+   * persistent storage, which must be retained (and not deleted) after the response has been sent. Therefore in the
+   * case of two identical delete requests being made, both will have access to slashing protection data.
+   *
+   * In a single atomic sequential operation the keymanager must:
+   * 1. Guarantee that key(s) can not produce any more signature; only then
+   * 2. Delete key(s) and serialize its associated slashing protection data
+   *
+   * DELETE should never return a 404 response, even if all pubkeys from request.pubkeys have no extant keystores
+   * nor slashing protection data.
+   *
+   * Slashing protection data must only be returned for keys from `request.pubkeys` for which a
+   * `deleted` or `not_active` status is returned.
+   *
+   * @param pubkeys List of public keys to delete.
+   * @returns Deletion status of all keys in `request.pubkeys` in the same order.
+   *
+   * https://github.com/ethereum/keymanager-APIs/blob/0c975dae2ac6053c8245ebdb6a9f27c2f114f407/keymanager-oapi.yaml
+   */
+  async deleteKeystores(
+    pubkeysHex: string[]
+  ): Promise<{
+    data: {
+      status: DeletionStatus;
+      message?: string;
+    }[];
+    slashingProtection: SlashingProtectionData;
+  }> {
+    const deletedKey: boolean[] = [];
+    const statuses = new Array<{status: DeletionStatus; message?: string}>(pubkeysHex.length);
+
+    for (let i = 0; i < pubkeysHex.length; i++) {
+      try {
+        const pubkeyHex = pubkeysHex[i];
+
+        // Skip unknown keys or remote signers
+        const signer = this.validator.validatorStore.getSigner(pubkeyHex);
+        if (!signer || signer?.type === SignerType.Remote) {
+          continue;
+        }
+
+        // Remove key from live local signer
+        deletedKey[i] = signer?.type === SignerType.Local && this.validator.validatorStore.removeSigner(pubkeyHex);
+
+        // Remove key from blockduties
+        // Remove from attestation duties
+        // Remove from Sync committee duties
+        // Remove from indices
+        this.validator.removeDutiesForKey(pubkeyHex);
+
+        // Remove key from persistent storage
+        for (const keystoreFile of await fs.promises.readdir(this.importKeystoresPath)) {
+          if (keystoreFile === `${KEY_IMPORTED_PREFIX}_${pubkeyHex}.json`) {
+            const keystorePath = join(this.importKeystoresPath, `key_imported_${pubkeyHex}.json`);
```

Comment:
Where are this keystore names define? You should define helper functions to ensure paths are the same in import / delete. Is this expecting 
```
`${KEY_IMPORTED_PREFIX}_${pubkeyHex}.json`
```
to be equal to 
```
join(this.importKeystoresPath, `key_imported_${pubkeyHex}.json`)
```
?

This logic feels very awkard

Reply:
Updated. Should be clearer now